### ISSUE TITLE
CIP3 concern: immutability in Bitcoin

### ISSUE THREAD

**Issue by jotapea** [CounterpartyXCP/cips#54]
I have concerns with CIP3, but I'm fairly new to Counterparty, so let me know if there is any misunderstanding in the following (and in the second part there are actual questions).

Following up on https://github.com/CounterpartyXCP/cips/issues/53, with CIP3 divisibility changes from an immutable asset property to a mutable one. Previously, the "genesis" asset issuance can always be used to know if an asset is divisible or not. But with CIP3, this would no longer be the case. And this is fundamentally relevant because divisibility affects the interpretation of quantities (divide by 10^8 or not).

One calculates the total quantity of an asset as:

```
total_quantity = sum_issuances - sum_destructions
```

With CIP3 the sums of these issuances/destructions will always need to know the 'latest asset reset' issuance.

And all this is fine in that it is totally doable. BUT now one of the most fundamental properties of an asset, its quantity, has been made more complex to code. AND what I believe is even more problematic, it has been made more complex to UNDERSTAND.

"Wait, asset quantities can be reset? Why is there a reset here in the total quantity calculation?"

I believe simplicity is essential in everything Bitcoin related. Adding resetting to something that is supposed to be immutable can be argued as a making it worse...

...

~~(2nd part)~~

~~While writing this I realize... will this be implemented in a way that makes all previous issuances/destructions invalid? If so, then the total quantity sum does not need to know about the 'latest asset reset' directly, as this is abstracted by only using valid issuances/destructions in sums.~~

~~Is this the case? I was wondering why there was a recent 'timestamp' while reading the messages from blocks... is it for changes like this?~~

~~But then, this means that Counterparty can make something that was previously valid, invalid?~~

Anyway to conclude, in Bitcoin, I believe it is debatable that CHANGING IMMUTABILITY can be considered an improvement. Where is this debate?

**Comment by jdogresorg** [comment]
1. Nowhere in Counterparty did anyone ever say that asset quantities were immutable and would never change, this is an incorrect assumption (unless you can point me to the comment your referencing where this was stated by a dev). CIP03 has been widely accepted by the community and it has been pending review for many 6+ years.

2. This CIP only effects asset owners who are in control of 100% of the supply, so it DOES NOT effect anyone who is an individual token holder besides the asset owner.

3. Every project that uses Counterparty can use the Counterparty API or XChain APIs to request current supply data,

https://docs.counterparty.io/docs/develop/api#get_supply
https://xchain.io/api#asset

As this is already the standard, I fail to understand why it is an issue that you can not determine supply using your above formula... unless your parsing in all CP transactions and keeping a state on your own, versus just using Counterparty for what it is for, getting state info and generating transactions.... are you doing this? is anyone? probably not.

4. Transactions on Counterparty are either valid or invalid at the time they are parsed by Counterparty, and the state is not changed after the fact. So, your concern about previous issuances/destructions being invalid is a red herring. All previous transactions retain their state (valid or invalid).

Allowing an asset owner to change asset divisibility when they control 100% of the supply DOES NOT make previous issuances/destroys invalid, they are still valid, and can be tracked in CP, as can the destruction of any existing supply at the time of reset, and the issuance of new supply. 

All state data is maintained in Counterparty, nothing changes with this CIP except the asset owner is able to change divisibility on their unused asset name/tokens, to make them actually usable.

**Comment by jotapea** [comment]
What is the purpose of the `balances` table? It seems to be a de-normalized view of the data already present in other tables, focused on efficiently returning the balance of an asset for an address.

https://github.com/CounterpartyXCP/counterparty-lib/blob/a99ddefd1a37751c9a090de63b3c106b82cb0bad/counterpartylib/lib/util.py#L460
```
sql='insert into balances values(:address, :asset, :quantity)'
```

But in terms of determining the TOTAL issuance of an asset, the source data are `issuances` and `destructions`. And the fundamental data model for both of these (and everything in Counterparty) are messages in blocks.

We all know Bitcoin's philosophy of "don't trust, verify". For anyone who follows this, they shouldn't trust an api for total quantity of an asset. They should read it directly from the messages in blocks. And currently, this works with the following simple calculation (which I will repeat, because changing this should not be taken lightly).

```
total_quantity = sum_issuances - sum_destructions
```

CIP3 breaks this. Because it is being considered from the perspective of the balance of an address. This is its current validation:

https://github.com/pataegrillo/counterparty-lib/blob/2ecf0ce129b0dd88057781619942ccefeca3b663/counterpartylib/lib/messages/issuance.py#L265
```

    if util.enabled("cip03", block_index) and reset and issuances:
        cursor = db.cursor()
        #Checking that all supply are held by the owner of the asset
        cursor.execute('''SELECT * FROM balances \
                            WHERE asset = ? AND quantity > 0''', (asset,))
        balances = cursor.fetchall()
        cursor.close()
        
        if (len(balances) == 0):
            if util.asset_supply(db, asset) > 0:
                problems.append('Cannot reset an asset with no holder')
        elif (len(balances) > 1):
            problems.append('Cannot reset an asset with many holders')
        elif (len(balances) == 1):
            if (balances[0]['address'] != last_issuance["issuer"]):
                problems.append('Cannot reset an asset held by a different address than the owner')
```

I believe Counterparty should not be doing anything with data views. Efficiency with de-normalization should be delegated to the application layers. The Counterparty api data should come as direct from the source as possible, which in the case of asset quantity is from issuances and destructions.

When done like this, I see that changing divisibility is a hard-fork kind of change, as it breaks the BASIC calculation of asset quantities. Why complicate things?

**Comment by jdogresorg** [comment]
`timestamp` is one of the parameters written to the `bindings` data in the `messages` table. 

The documentation indicates what data fields you will get back from the `messages` table, it does not describe every possible field which might be described in the`bindings` field, as this could be any field from any table.

**Comment by jotapea** [comment]
I didn't expect this issue to be closed after that question, which was tangential. @jdogresorg can you reopen this issue to continue the discussion? I would like to know your response to the main comments here. And general feedback from ANYONE else.

I see in https://github.com/CounterpartyXCP/counterparty-lib/pull/1193#issuecomment-1212574763 that the CIP3 activation is set for block 753000.

If I am bringing up concerns with this Counterparty Improvement Proposal, because from my perspective this is NOT an Improvement, why not give this argument a chance? There are only about 3 weeks left until said block, and my hope is to find the best solution for Counterparty (and Bitcoin) long term, which is something we can all agree on.

> This CIP only effects asset owners who are in control of 100% of the supply, so it DOES NOT effect anyone who is an individual token holder besides the asset owner.

This is similar to to what is said in the [rationale](https://github.com/CounterpartyXCP/cips/blob/master/cip-0003.md#rationale) of CIP itself, but this is debatable. I believe Counterparty as a Bitcoin platform gets eternally affected by this change!

Assets are created by issuances. Issuances specify quantity and interpretation of this quantity, with divisibility. To know total quantity of how much has been issued for an asset, is just a sum of all issued quantities. And the interpretation of this quantity is based on the divisibility shared by ALL issuances. All issuance data for an asset is usable and required to know the total.

With CIP3, anyone who wants to build on Counterparty, will now need to know about this SPECIAL case where people can invalidate issuances, just to UNDO divisibility! What in Bitcoin is undo-able? And because of this special niche "fix-a-mistake" use case, the FUNDAMENTAL way to calculate asset quantity stops working. Having issuances with different interpretations of quantity makes all previous issuances before a reset UNUSABLE, wasted.

These rationales would still be valid if the person "by mistake" locked the asset. Why not allow undo-ing this also?

CIP3 may had well-intentions at its time. But nowadays, with solutions like subassets (which I am aware was proposed by you @jdogresorg and is great), the 'root' asset name has more purpose. And because we are talking about unlocked assets, you can still increase issuance or reduce it with destroys.

> to make them actually usable.

Talking about making them usable just by being able to change divisibility is not a valid argument for me. Any root asset can create infinite amount of subassets, each with any combination of parameters wanted.

There is a reason why this was never prioritized, why rush it now? I believe CIP3 only encourages name squatting and wasting the Bitcoin blockspace.

**Comment by jdogresorg** [comment]
I closed the issue because I believe that discussion on this topic will not change the outcome of CIP3, which will be implemented as stated. I will reopen the issue for further discussion.

>why not give this argument a chance? 

I am open to hearing what you have to say, but discussion has been open on this CIP for 8+ years, plenty of time for any objections to be raised and discussed previously.


> These rationales would still be valid if the person "by mistake" locked the asset. Why not allow undo-ing this also?

Because the previous developers have explicitly stated that once an asset is locked, it is locked forever, and can never be unlocked again. This is why we are not allowing 'undo-ing' locks. We have made no such statement about divisibility never being changed, and in fact have made intentions clear to allow divisibility changes for 8+ years while this CIP has been pending.

> Talking about making them usable just by being able to change divisibility is not a valid argument for me. 

I think we can agree to disagree on this point. I have assets which I have registered, almost all as divisible. I now have buyers who are interested in purchasing asset names from me, but they want the ability to change the token from a divisible token to a non-divisible one (so they can use as an NFT). Your solution would be to tell the users that they can't own MCDONALDS and use it how they want, but that they can just issue subassets instead, which IMO is not really a viable solution. Asset owners should be able to use the base asset however they would like, including resetting asset divisibility and supply in special cases.

> There is a reason why this was never prioritized, why rush it now?

It was never really prioritized before because there was not much of a secondary market for buying unused asset names. Now there are a decent number of users who are purchasing asset ownership and want the ability to reset the token supply from divisible to non-divisible (as most NFTs now are issued as non-divisible tokens)

We will be updating `get_balances` data to start passing a `divisible` field, so you can still easily determine divisibility without having to make any additional API calls.

**Comment by jotapea** [comment]
The "[main](https://counterpartytalk.org/t/cip-reset-token-divisibility-statuses-for-unused-asset/1643)" discussion for this CIP started in November 16, 2015. This discussion ends in December 8, 2015. After the last comment from December 8, no other message is discussing the CIP. There are 3 more comments from November 2017 which end up concluding these are another topic. And then the last one from February 13, 2019: "lock your assets, you can keep your assets". So, no discussion on the topic for almost the entirety of the last <7 years.

Could it be that because of the existence of CIP3, Counterparty user behavior was affected? Is it because there was always hope that if these squatted assets could be reset to be resold, by these specific parameters (creator keeps 100% of the supply), that the motivation to use these diminished? Was having thousands of unused good names, for many years, good for the project? It doesn't feel like it, I think Counterparty has become relevant mainly because of the NFT boom, not because of names.

Which by the way, NO ASSET NAME CAN START WITH A! This is a big limitation that won't allow these to become as mainstream as domain names for example. But I actually LOVE IT, because it is awesome that A Assets are the free (as in no xcp needed) btc-only ones!

I'm learning that CIP3 is literally focused on the OG asset name squatters, which puts me as a newer member of the community in a conflicting position with the OGs... but the thing is, I INTEND to become another Counterparty OG! I'm already almost halfway there with Bitcoin (started in 2017).

I believe Counterparty needs to be more strategic as it moves forward. CIP3, in its current form (and as the "future plan" it has been), feels like an unhealthy feature addition, as it incentivizes and rewards NOT using assets. And even if it has backing by some of the TOP community contributors, people I FULLY respect, we all have to realize they are also individuals with their own motivations (just like I have mine).

If something like CIP3 wants to be done, it needs to be done in a way that adds general value for everyone in the community. If Counterparty grows up to its potential, THE NEW PEOPLE WILL BECOME THE VAST MAJORITY. These new people have no such attachment to asset names, as the most obvious resellable brand-like names are out of the picture.

If the fundamental way of calculating assets
```
total_quantity = sum_issuances - sum_destructions
```
is going to be broken forever, then it must be worth it based on the value added to the majority of the community.

These are general feedback I have received or suggested:

- allow assets to be unlocked within the same rationale as CIP3
- allow changes in divisibility no matter what before it is locked
- completely remove the option of divisibility
- treat all unlocked assets at satoshi level (integers) and make divisibility a locked-only property

Are there other alternatives to consider?

**Comment by jdogresorg** [comment]
> So, no discussion on the topic for almost the entirety of the last <7 years.

To clarify, your assumption is incorrect, there have been numerous discussions of CIP in slack and telegram over the past few years, simply because all conversation is not displayed on the forum threads DOES NOT mean that there has not been conversation.

> Was having thousands of unused good names, for many years, good for the project? It doesn't feel like it, 

I would argue YES, as that is the primary reason why xchain.io (the only block explorer) and freewallet.io (the main wallet) exist, because I had a sunken cost in asset names and a vested interest in supporting the platform... Had I not invested funds in registering thousands of asset names, I probably would not be as dedicated to Counterparty over the years as I have been.

> Which by the way, NO ASSET NAME CAN START WITH A! This is a big limitation that won't allow these to become as mainstream

And yet, the platform has continued to operated perfectly fine for 8+ years without asset names that start with A, and has seen decent adoption, so again, we can agree to disagree on this point.

> I'm learning that CIP3 is literally focused on the OG asset name squatters

This is incorrect, you may get this assumption because I am supportive of CIP3, but CIP3 was written by JPJ many years ago (and he isn't a huge squatter of names) to benefit all users of counterparty that have unused asset names. Please stop trying to paint the incorrect picture that CIP3 is written and being pushed for by big squatters, as that is not the case.

**Comment by jdogresorg** [comment]
> So, no discussion on the topic for almost the entirety of the last <7 years.

To clarify, your assumption is incorrect, there have been numerous discussions of CIP in slack and telegram over the past few years, simply because all conversation is not captured or displayed on the forum threads DOES NOT mean that there has not been conversations in the past 7+ years.

> Was having thousands of unused good names, for many years, good for the project? It doesn't feel like it, 

I would argue YES, as that is the primary reason why xchain.io (the only block explorer) and freewallet.io (the main wallet) exist.... because I had a sunken cost in asset names and a vested interest in supporting the platform... Had I not invested funds in registering thousands of asset names, I probably would not be as dedicated to Counterparty over the years as I have been. So, while you might not see the value in name squatting, I can assure you that it has benefitted the ecosystem greatly.

> Which by the way, NO ASSET NAME CAN START WITH A! This is a big limitation that won't allow these to become as mainstream

And yet, the platform has continued to operated perfectly fine for 8+ years without asset names that start with A, and has seen decent adoption, so again, we can agree to disagree on this point.

> I'm learning that CIP3 is literally focused on the OG asset name squatters

This is incorrect, you may get to this assumption because I am supportive of CIP3, but CIP3 was written by JP many years ago (and he isn't a huge squatter of names) to benefit all users of Counterparty that have unused asset names. If this was a CIP3 that was aimed at OG asset name squatters, it could have been pushed through a long time ago. Please stop trying to paint the incorrect picture that CIP3 is written and being pushed for by big squatters, as that is not the case. CIP03 was left open for many years until there was decent demand for the feature, and is only now getting added to Counterparty because there is an actual demand to be able to reset asset supply/divisibility.

>If something like CIP3 wants to be done, it needs to be done in a way that adds general value for everyone in the community.

That is exactly what is happening here, it is unfortunate that you do not see being able to recycle tens of thousands of assets back into the ecosystem as a net win, all because you can't stick with your formula that you continue to repeat (seriously, 3x times repeating `total_quantity = sum_issuances - sum_destructions` is enough... we your point, no need to keep repeating it)

The value add of having thousands of assets recycled back into the ecosystem, and allowing asset owners to control their asset entirely, including reseting divisibility when they are in control of 100% of the supply, is a net win for the Counterparty community, and it is unfortunate that you do not share this viewpoint. Perhaps engaging with the CP community a bit more and hearing feedback from actual end users would help shape a more informed viewpoint.

At this point, I feel I have spent ample time trying to explain how CIP3 works, and why it benefits the community. I have heard your concerns about divisibility, and we can agree to disagree that this is a net benefit for Counterparty users.

**Comment by jotapea** [comment]
Made this [PR](https://github.com/CounterpartyXCP/counterparty-lib/pull/1201).

**Comment by jotapea** [comment]
> Which by the way, NO ASSET NAME CAN START WITH A! This is a big limitation that won't allow these to become as mainstream as domain names for example. But I actually LOVE IT, because it is awesome that A Assets are the free (as in no xcp needed) btc-only ones!

To continue with the comparison with domain names:

- domain names can never be lost forever (in Counterparty you can loose your keys)
- domain names have no LOCK, after a domain was used by one entity it can then be used by a completely different entity in the future without any limitation set by prior owners (in CP the first lock makes it forever)

And these limitations are not necessarily bad, it just makes asset names serve a different purpose than domain names. These differences show that asset names should not be given so much importance as CIP3 is giving them. Their main purpose is being customizable unique ids, not brands or words.

> FYI, your formula total_quantity = sum_issuances - sum_destructions still works to determine the total quantity even after CIP03, since before an asset can be reset its existing supply will be destroyed via destructions.

After learning that the reset includes making a destroy of what is currently in circulation, and making some test calculations, it does make the total asset quantity calculation consistent. So this is accurate.

BUT then I still have to focus on how for this niche use-case, it affects the integrity of the individual calculations of `sum_issuances` and `sum_destructions`. Before CIP3 you can always understand the sum of issuances/destructions as the number can only be interpreted in one way (either divisible or indivisible) based on the genesis issuance.

Some could consider divisibility as a Counterparty design flaw. With CIP3, we are doubling down on this flaw, adding complexity and affecting data integrity, with no real need.

Why no need? If anyone wants to make a divisible asset as an NFT, why not do the exact same destroys that the reset would do, minus one. Satoshis are indivisible.

The following are the [example use cases](https://github.com/CounterpartyXCP/cips/blob/master/cip-0003.md#example-use-cases) from CIP3:

> Example use cases
>
> Alice owns GOLD. The supply is 88,888,888 shares. Bob is setting up a GOLD ETF on Counterparty and intends to let one share represent 1 oz of gold. The ETF's total supply is only 10,000 oz. Alice wants to sell GOLD to Bob, and Bob would have been interested if he could set the supply to 10,000 shares.
>
> Carol registered ODIN long ago. Now she plans to make a card game with ancient gods. Players are encouraged to trade on the DEX but ODIN is divisible and cards should not be that. If someone puts an ODIN card for sale, a troll may match the order with a millionth of a card, hence destroying the card (this actually happened with a Spells of Genesis card)

Literally 1 of the 2 use cases has already been 100% solved with the introduction of destroys. And interestingly, the other one can also be solved with destroys, by destroying your supply down to the satoshi level and treating these satoshis as your indivisible units.

> Are there other alternatives to consider?

I missed the most important alternative here: DO NOTHING! Lets just focus on improving the UI/UX of the applications. Lets make divisibility default to false (opposite from the [api](https://counterparty.io/docs/api/#create_issuance)) if the top use case for assets are NFTs. Lets do user interfaces that treat satoshi level NFTs the same as non-divisible, which btw, this is [how it is done](https://counterparty.io/docs/api/#quantities-and-balances) at the protocol level.

CIP3 Pros (only for non-locked 100% supply issuer-controlled assets):
- change divisibility (until locked or distributed)
- "improved" resell-ability (until locked or distributed)

CIP3 Cons:
- complicates the protocol
- affects data integrity
- not needed if divisible use destroys to be treated at the satoshi level
- not needed if indivisible use issuances to be as divisible as having 10^8 more units
- doubles down on what some could consider a CP design flaw
- incentivizes and rewards not using assets
- minimizes the privilege that root assets have to create infinite subassets with any parameters
- wastes Bitcoin blockspace
- goes AGAINST Bitcoin's immutability principles

My conclusion is that CIP3, as it was originally proposed, should not be done. Counterparty as it stands today can almost fully make up for it, so the negatives added are not worth it. Change divisibility by burning to satoshi level or issuing to 10^8 more units, then resell as you wish.

**Comment by loon3** [comment]
While I appreciate @jotapea's thoughtful comments, they are primarily opinions about how things should be rather than the reality of how things are.  Calling out "design flaws" and "wasting Bitcoin blockspace" is also not helpful as it is vague and could be leveraged against any protocol improvement, even Counterparty itself.  Alternative suggestions to CIP3 like "destroying your supply down to the satoshi level and treating these satoshis as your indivisible units" or the inverse for non-divisible assets are actually creating more complexity at the client level because now those assets must be treated differently across platforms with no clear signal of how they should be treated coming from the Counterparty node.  

Counterparty is a social protocol and any introduction of a new message type or new parameters to existing message types are nearly always a hard fork.  Dispensers and destructions are both great examples of this.  There have been many hard forks over the years [see Protocol Changes](https://github.com/CounterpartyXCP/counterparty-lib/blob/master/counterpartylib/protocol_changes.json)

I also take issue with CIP3 going "AGAINST Bitcoin's immutability principles".  Counterparty assets are not bitcoin.  They are user issued and as such the issuer is free to change them within the parameters available.  Issuers have the ability to increase supply of their unlocked assets whether distributed or not.  CIP3 goes a step further and requires that assets not be distributed virtually eliminating any risk of a client wallet for misinterpreting the divisibility status of the asset.  Once an asset is locked no further supply can be issued and no further divisibility changes can be made.  CIP3 does not make any changes to the existing social contract.

I do think you make a reasonable point regarding determining total issuance based on the `issuances - destructions`.  My suggestion to make the CIP3 divisibility reset as clean as possible is for that particular message to also result in the total destruction of asset supply.  This way a divisibility reset also means an asset supply reset.  The reset issuance tx can be thought of as the new "genesis" issuance, effectively a new epoch for that asset with all subsequent issuance txs determining supply.  I believe this change would satisfy your initial and primary concern.

**Comment by droplister** [comment]
Make the activation block further in the future. Counterparty is a small, federated network. This is a state management change that needs more time than two weeks to accommodate, though it can be accommodated for.

**Comment by jotapea** [comment]
Thank you for your input @loon3, also helpful in understanding more about Counterparty.

My specific concern with the `issuances` and `destructions` is now about the individual sums of these. Before CIP3, all issuances add up to a number that makes sense. And all destroys also add up to a number that makes sense. And their difference is the actual quantity. After CIP3, ONLY this difference keeps consistency (as long as you use the correct divisibility interpretation), but no longer their individual sums.

But I have no problem with the feasibility of CIP3. My main problem is that it adds an unnecessary layer of complexity that affects the philosophical principles of this Bitcoin project. I don't see Counterparty separate from Bitcoin, as it exists inside of Bitcoin.

Anyone trying to understand and work with the Counterparty protocol will need to add understanding this "reset" as the reason why these sums no longer make sense ONLY for a specific set of assets ONLY because, under certain specific conditions, divisibility can be changed. And just because of this, add processing, energy, and cost for all Counterparty nodes (and developers) for the rest of time.

CIP3 breaks the simple cohesive elegant design of issuances and destructions.

I have some questions:

- Can this reset be done multiple times?
- What happens with the descriptions done before a reset?

Issuances contain quantity and description in a message. This reset cannot erase history, every event done by an asset is recorded forever. There is only a single genesis, and an asset's history is inherently part of it.

Divisibility is the ONLY property that is final from genesis. Similar to locking maximum supply is forever, I see the divisibility choice from the first issuance as a "lock of intent". Is Counterparty self-preserving. I like that there is still ONE aspect of this unique name reservation system that is eternal, but CIP3 is taking it away.

**Comment by jdogresorg** [comment]
> "Can this reset be done multiple times?"

Yes, a token supply can be reset many times as long as the supply has never been LOCKED and the asset owner is in control of 100% of the supply

> "What happens with the descriptions done before a reset?

The description remains the exact same as it was before a reset. A reset only effects resetting the supply, not the asset description.

**Comment by jotapea** [comment]
> Yes, a token supply can be reset many times as long as the supply has never been LOCKED and the asset owner is in control of 100% of the supply

This makes more sense than only once, as that would just move the immutability one step. BUT this means that Counterparty software must always be aware of the latest reset... until locked. This is different from locks as these are always about the first time (for some reason Counterparty allows re-locking assets, why is this?).

And in this aspect is like destroys, which could allow someone to continuously issue and destroy assets indefinitely. So the same could be done with resets, BUT definitely more complex, as the interpretation of quantities can mean anything based on the reset that applies at that time. There is a reason why divisibility is not an included property of a destroy message.

> The description remains the exact same as it was before a reset. A reset only effects resetting the supply, not the asset description.

Then this one surfaces a big contradiction to the reset concept. So now the reset LITERALLY only means reset of divisibility, as all previous issuances must still be considered for their "partial validity" with their descriptions. This DEFINITELY means the reset cannot be considered an new epoch for an asset.

**Comment by loon3** [comment]
> BUT this means that Counterparty software must always be aware of the latest reset... until locked.

This is also true for total issuance plus keep in mind this has zero effect on any user beyond the issuer themselves because they must hold total issuance to reset.

> Then this one surfaces a big contradiction to the reset concept. So now the reset LITERALLY only means reset of divisibility, as all previous issuances must still be considered for their "partial validity" with their descriptions. This DEFINITELY means the reset cannot be considered an new epoch for an asset.

Descriptions are ALWAYS mutable even when an asset is locked.  There's no expectation that they won't change block to block.  The "problem" you are describing will not change under CIP3.

**Comment by jotapea** [comment]
Are descriptions really mutable? Or is just the UI/UX showing the latest description? All past descriptions are part of the asset's history. Descriptions are added, not replaced.

**Comment by loon3** [comment]
You just described a blockchain.

**Comment by jdogresorg** [comment]
Closing issue as CIP3 code has been merged and released in v9.60.0
